libsel4vmmplatsupport/src/arch/arm/guest_image.c

```diff
@@ -148,48 +155,47 @@ static int guest_write_address(vm_t *vm, uintptr_t paddr, void *vaddr, size_t si
     return 0;
 }
 
+static size_t get_file_size(const char *image_name)
+{
+    FILE *file = fopen(image_name, "r");
+    if (!file) {
+        return 0;
+    }
+    fseek(file, 0, SEEK_END);
+    size_t size = ftell(file);
+    fseek(file, 0, SEEK_SET);
+
+    fclose(file);
+
+    return size;
+}
+
 static int load_image(vm_t *vm, const char *image_name, uintptr_t load_addr,  size_t *resulting_image_size)
 {
     int fd;
-    size_t len;
     int error;
+
+    size_t file_size = get_file_size(image_name);
```

Comment:
Do we need this helper of can we use `stat(fd, &stat_data)` and then get the file size from `stat_data.st_size`?

Reply:
That doesn't work.

This block:
```
    struct stat buf;
    stat(image_name, &buf);

    ZF_LOGE("%d | %d", file_size, buf.st_size);
```

Gives this output:
```
libsel4muslcsys: Error attempting syscall 79
load_image@guest_image.c:195 15405568 | 0
Loading Initrd: 'linux-initrd'
libsel4muslcsys: Error attempting syscall 79
load_image@guest_image.c:195 912857 | 0
```

Reply:
That doesn't work either.

I can get things working w/o the helper function by using `fread`:

```
static int guest_write_address(vm_t *vm, uintptr_t paddr, void *vaddr, size_t size, size_t offset, void *cookie)
{
    FILE *file = (FILE *)cookie;

    /* Load the image */
    int err = fread(vaddr, size, 1, file);
    if (err != 1) {
        ZF_LOGE("fread failed unexpectedly");
        return -1;
    }

    if (config_set(CONFIG_PLAT_TX1) || config_set(CONFIG_PLAT_TX2)) {
        seL4_CPtr cap = vspace_get_cap(&vm->mem.vmm_vspace, vaddr);
        if (cap == seL4_CapNull) {
            /* Not sure how we would get here, something has gone pretty wrong */
            ZF_LOGE("Failed to get vmm cap for vaddr: %p", vaddr);
            return -1;
        }
        int error = seL4_ARM_Page_CleanInvalidate_Data(cap, 0, PAGE_SIZE_4K);
        ZF_LOGF_IFERR(error, "seL4_ARM_Page_CleanInvalidate_Data failed");
    }
    return 0;
}

static int load_image(vm_t *vm, const char *image_name, uintptr_t load_addr,  size_t *resulting_image_size)
{
    int error;

    FILE *file = fopen(image_name, "r");
    if (!file) {
        ZF_LOGE("Error: Unable to find image \'%s\'", image_name);
        return -1;
    }

    fseek(file, 0, SEEK_END);
    size_t file_size = ftell(file);
    fseek(file, 0, SEEK_SET);

    if (0 == file_size) {
        ZF_LOGE("Error: \'%s\' has zero size", image_name);
        return -1;
    }

    vm_ram_mark_allocated(vm, load_addr, ROUND_UP(file_size, PAGE_SIZE_4K));
    error = vm_ram_touch(vm, load_addr, file_size, guest_write_address, (void *)file);
    if (error) {
        ZF_LOGE("Error: Failed to load \'%s\'", image_name);
        fclose(file);
        return -1;
    }

    *resulting_image_size = file_size;
    fclose(file);
    return 0;
}

@axel-h @kent-mcleod do you guys like that method better? Or should we stick with using the file descriptor / open method?
```

Reply:
On a naive view, switching from the bare metal `open()` to the more file specific `fopen()` would be a generic refactoring that could happen first, as it seems this is the proper API-familiy to use for file I/O.

Reply:
In that case I'll push the `fopen` version. Using the `file` pointer as a cookie is also more straight forward.

Reply:
> OK, thanks for trying. Out of curiosity, does this work with the file descriptor
> 
> ```
>     seek(fd, 0, SEEK_END);
>     size_t size = tell(fd);
>     seek(fd, 0, SEEK_SET);
> ```

@chrisguikema, what was the issue with this approach? `lseek(fd, 0, SEEK_END);` should directly return the filesize in the return value. Then `lseek(fd, 0, SEEK_SET);` should reset it. This is what the fseek and ftell functions are using internally.

It's probably unsafe to use the file streaming layer because muslc assumes extra OS functionality that isn't implemented by sel4muslcsys. It also adds an extra layer of buffering which we don't use.

Reply:
```
    size_t size = lseek(fd, 0, SEEK_END);
    lseek(fd, 0, SEEK_SET);
```

Using `lseek` works. When using `seek` and `tell` there were linker issues. I can update to not use the helper in this case.